File: LeetCode/DP/Python/Leetcode813-largest-sum-of-averages.py
class Solution:
    def largestSumOfAverages(self, nums: List[int], k: int) -> float:
        n = len(nums)
        s, f = [0] * (n + 1), [[-1000000] * (k + 1) for _ in range(n + 1)]
        for i in range(1, n + 1): s[i] = s[i - 1] + nums[i - 1]
        f[0][0] = 0
        for i in range(1, n + 1):
            for j in range(1, k + 1):
                for q in range(i):
                    f[i][j] = max(f[i][j], f[q][j - 1] + (s[i] - s[q]) / (i - q))
        return f[n][k]

File: LeetCode/DP/Rust/Leetcode813-largest-sum-of-averages.rs
impl Solution {
  pub fn largest_sum_of_averages(nums: Vec<i32>, k: i32) -> f64 {
    let n = nums.len();
    let (mut f, mut s) = (vec![vec![-10000f64; k as usize + 1]; n + 1], vec![0f64; n + 1]);
    f[0][0] = 0f64;
    for i in 1..=n {
      s[i] = s[i - 1] + nums[i - 1] as f64;
    }
    for i in 1..=n {
      for j in 1..=k as usize {
        for q in 0..i {
          f[i][j] = f[i][j].max(f[q][j - 1] + (s[i] - s[q]) / (i as f64 - q as f64));
        }
      }
    }
    f[n][k as usize]
  }
}
